sdk/examples/xo_python/sawtooth_xo/processor/xo_state.py

```diff
@@ -0,0 +1,160 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+import hashlib
+
+from sawtooth_sdk.processor.exceptions import InternalError
+
+
+XO_NAMESPACE = hashlib.sha512('xo'.encode("utf-8")).hexdigest()[0:6]
+
+
+def _make_xo_address(name):
+    return XO_NAMESPACE + \
+        hashlib.sha512(name.encode('utf-8')).hexdigest()[:64]
+
+
+class Game(object):
+    def __init__(self, name, board, state, player1, player2):
+        self.name = name
+        self.board = board
+        self.state = state
+        self.player1 = player1
+        self.player2 = player2
+
+
+class XoState(object):
+
+    TIMEOUT = 3
+
+    def __init__(self, context):
+        """Constructor.
+
+        Args:
+            context (sawtooth_sdk.processor.context.Context): Access to
+                validator state from within the transaction processor.
+        """
+
+        self._context = context
+        self._games = None
+
+    def delete_game(self, game_name):
+        """Delete the Game named game_name from state.
+
+        Args:
+            game_name (str): The name.
+        """
+
+        del self._games[game_name]
+        if self._games:
+            self._store_game(game_name)
+        else:
+            self._delete_game(game_name)
+
+    def set_game(self, game_name, game):
+        """Store the game in the validator state.
+
+        Args:
+            game_name (str): The name.
+            game (Game): The information specifying the current game.
+        """
+
+        self._games[game_name] = game
+
+        self._store_game(game_name)
+
+    def get_game(self, game_name):
+        """Get the game associated with game_name.
+
+        Args:
+            game_name (str): The name.
+
+        Returns:
+            (Game): All the information specifying a game.
+        """
+
+        return self._load_game(game_name=game_name)
+
+    def _store_game(self, game_name):
+        address = _make_xo_address(game_name)
+
+        state_data = self._serialize(self._games)
+
+        self._context.set_state(
+            {address: state_data},
+            timeout=self.TIMEOUT)
+
+    def _delete_game(self, game_name):
+        address = _make_xo_address(game_name)
+        self._context.delete_state(
+            [address],
+            timeout=self.TIMEOUT)
+
+    def _load_game(self, game_name):
+        address = _make_xo_address(game_name)
+
+        state_entries = self._context.get_state(
+            [address],
+            timeout=self.TIMEOUT)
+        if state_entries:
+
+            games = self._deserialize(data=state_entries[0].data)
+
+        else:
+            games = {}
+
+        self._games = games
```

Comment:
Per our discussion, I think the `XoState` class should be a thin wrapper around the context object that handles (de)serialization of games, hash collisions, and address creation. In this model, it makes sense to maintain an up-to-date cache between the handler and the validator to reduce the number of messages that need to be exchanges and keep track of any games at the address of interest that are not the game of interest because of a hash collision.

I think another valid API, which it sounds like you started with, would be to have one XoState object per game you are interacting with. However in that case, you would probably want the class to be called XoGame and to take the game name as a constructor argument. In that case, the class is more of a network storage backed RPC object.

While both are valid, it looks like we are already using the first one in other places such as intkey rust: https://github.com/hyperledger/sawtooth-core/blob/master/sdk/examples/intkey_rust/src/handler.rs#L139